Allow for running AWX at non-root path (URL prefixing)

shanemcd opened the pull request:
<!--- changelog-entry
# Fill in 'msg' below to have an entry automatically added to the next release changelog.
# Leaving 'msg' blank will not generate a changelog entry for this PR.
# Please ensure this is a simple (and readable) one-line string.
---
msg: "Added support for running AWX at non-root path (URL prefixing)"
-->

##### SUMMARY
These changes allow for running AWX at a non-root path, like `https://example.com/awx/`:

![Screen Shot 2021-11-14 at 12 38 15 PM](https://user-images.githubusercontent.com/773186/141667765-b1cb3e30-82ed-4795-a0b0-62880cc3cf20.png)

Must be merged along with https://github.com/ansible/awx-operator/pull/652

shanemcd committed:
Move production UWSGI config to a file

shanemcd committed:
Allow for overriding UWSGI mount path

This is just one piece of the puzzle as I try to add support for URL prefixing.

shanemcd committed:
Use relative paths for UI assets

Found at https://create-react-app.dev/docs/deployment/#serving-the-same-build-from-different-paths

shanemcd committed:
Make UI work when not running at root path

shanemcd committed:
Fix bug in named url middleware when running at non-root path

The most notable change here is the removal of the conditional in
process_request. I don't know why we were preferring REQUEST_URI over
PATH_INFO. When the app is running at /, they are always the same as far as I
can tell. However, when using SCRIPT_NAME, this was incorrectly setting path and
path_info to /myprefix/myprefix/.

shanemcd committed:
Bind port for UI live reload tooling in development environmentt

This allows for running:

```
docker exec -ti tools_awx_1 npm --prefix=awx/ui start
```

shanemcd commented:
Added instructions on how to test at https://github.com/ansible/awx-operator/pull/652#issue-780017026

pabelanger commented:
\o/

sean-m-sullivan commented:
So I looked at this, and it looks brilliant, but I wonder how its going to affect the awx.awx collection, does it effect how the url is generated there?

AlanCoding commented:
If a client does splitting on the `/` characters, it should split starting from the right. However, we still have testing we should still run for this.

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -390,7 +391,20 @@ def get_related(self, obj):
+
+            # If the app is running at a location other than /, temporarily remove the
+            # prefix. This is to avoid changing the code in _generate_named_url where
+            # it is assumed the ID of the resource is at url_units[4].
```
this seems like a hacky way to solve the problem, as opposed to modifying `_generate_named_url` itself.

shanemcd replied:
Suggestions? We don't have access to the request object there. Sending the index in as an argument seems equally weird to me.

shanemcd replied:
@AlanCoding Please take a look at my latest commit.

AlanCoding commented on tools/ansible/roles/dockerfile/templates/supervisor.conf.j2:
```diff
@@ -28,7 +28,7 @@ directory = /awx_devel
     UWSGI_DEV_RELOAD_COMMAND='supervisorctl -c /etc/supervisord_task.conf restart all; supervisorctl restart tower-processes:daphne tower-processes:wsbroadcast'
 {% else %}
-command = /var/lib/awx/venv/awx/bin/uwsgi --socket 127.0.0.1:8050 --module=awx.wsgi:application --vacuum --processes=5 --harakiri=120 --no-orphans --master --max-requests=1000 --master-fifo=/var/lib/awx/awxfifo --lazy-apps -b 32768
+command = /var/lib/awx/venv/awx/bin/uwsgi /etc/tower/uwsgi.ini
```
oh this is very nice

shanemcd committed:
Simplify code that generates named URLS

rooftopcellist commented:
It looks like removing this import is what is needed to make the linters happy:

```
awx/api/serializers.py:100:1: F401 'awx.main.utils.removeprefix' imported but unused
```

jakemcdermott commented on tools/docker-compose/ansible/roles/sources/templates/docker-compose.yml.j2:
```diff
@@ -56,6 +56,7 @@ services:
       - "8013:8013"  # http
       - "8043:8043"  # https
       - "2222:2222"  # receptor foo node
+      - "3001:3001"  # used by the UI dev env
```
```suggestion
      - "3000:3001"  # used by the UI dev env
```

shanemcd committed:
Update port binding for UI dev tooling

Jake says "Folks sometimes run the ui dev server independently of the tools_awx container"

Co-authored-by: Jake McDermott <9753817+jakemcdermott@users.noreply.github.com>

adpavlov commented:
It seems that Mesh.js also needs to be changed to remove leading / in

`this.baseUrl = '/api/v2/mesh_visualizer/';`

@shanemcd
